src/lib/slot.c

```diff
@@ -139,8 +144,23 @@ CK_RV slot_mechanism_list_get (CK_SLOT_ID slot_id, CK_MECHANISM_TYPE *mechanism_
         return CKR_BUFFER_TOO_SMALL;
     }
 
-    *count = ARRAY_LEN(mechs);
-    memcpy(mechanism_list, mechs, sizeof(mechs));
+    TPMS_CAPABILITY_DATA *capabilityData = NULL;
+    rv = tpm_get_algorithms (t->tctx, &capabilityData);
```

Comment:
I wonder if this would be better on C_Init() and then cache it.

Reply:
The general question is, what shall we cache? (in the context?)
The algorithms capability?, the supported mechanism?, the mechanism+details?

Reply:
The overall current approach is awkward even if you don't just cache the data structures on init and do it on demand. None of these structures then would be a static list you filter non-supported things out of. Just always ask the TPM and return the values. If someone doesn't provide size, ask the TPM and return the list count. Add caching for bonus points.